core/src/main/scala/org/apache/spark/eventhubs/PartitionsStatusTracker.scala

```diff
@@ -0,0 +1,379 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.eventhubs
+
+import java.util.logging.Logger
+
+import scala.collection.mutable
+import scala.collection.breakOut
+import org.apache.spark.eventhubs.rdd.OffsetRange
+import org.apache.spark.eventhubs.utils.ThrottlingStatusPlugin
+import org.apache.spark.internal.Logging
+
+
+class PartitionsStatusTracker extends Logging{
+
+  import PartitionsStatusTracker._
+
+  // retrives the batchStatus object based on the local batchId
+  private val batchesStatus = mutable.Map[Long, BatchStatus]()
+
+  /**
+   * retirives the local batchId for a pair of (NameAndPartition, SequenceNumber)
+   * it's useful to accss the right batch when a performance metric message is
+   * received for a parition-RequestSeqNo pair.
+   * it's getting updated every time a batch is removed or added to the tracker
+   */
+  private val partitionSeqNoPairToBatchIdMap = mutable.Map[String, Long]()
+
+  /**
+   * Add a batch to the tracker by creating a BatchStatus object and adding it to the map.
+   * Also, we add mappings from each partition-startSeqNo pair to the batchId in order to be able to retrive
+   * the batchStatus object from the map when the performance metric message is received.
+   * Not that we ignore partitions with batchSize zero (startSeqNo == latestSeqNo) since we won't recieve
+   * any performance metric message for such partitions.
+   */
+  def addBatch(batchId: Long,
+               offsetRanges: Array[OffsetRange]): Unit = {
+    if(batchesStatus.contains(batchId)) {
+      throw new IllegalStateException(
+        s"Batch with local batch id = $batchId already exists in the partition status tracker. Can't be added again.")
```

Comment:
Do we expect the same batchId will be added more than once? What about update the batchStatus with the latest data in that case instead of throwing an error? I.e., AddOrUpdate(..)

Reply:
Batches are not supposed to be updated. However, I changed the API as you suggested and added a log entry if a batch is updated instead of an exception.